app/features/build_runner/remote_runner.rb

```diff
@@ -0,0 +1,330 @@
+      )
+      responses.each do |response|
+        validate_agent_response(response: response)
+        process_agent_response(response: response)
+      end
+      emit_new_row(type: :last_message, message: nil, time: Time.now)
+
+      current_build.duration = Time.now.utc - start_time
+      if @build_state.state == :SUCCEEDED
+        current_build.status = :success
+        current_build.description = "All green"
+        logger.info("fastlane run complete")
+      else
+        current_build.status = :failure
+      end
+
+      @current_build.artifacts.each do |artifact|
+        project.artifact_provider.store!(artifact: artifact, build: current_build, project: project)
+      end
+
+      save_build_status!
+      Services.build_runner_service.remove_build_runner(build_runner: self)
+      return @build_state.state == :SUCCEEDED
+    end
+
+    def validate_agent_response(response:)
+      if @build_state.state == :PENDING && !response.state
+        raise "First InvocationResponse should change the state from :PENDING"
+      end
+      if response.state != :PENDING
+        unless @build_state.trigger?(response.state)
+          raise "Unexpected state change #{@build_state.state} -> #{response.state}"
+        end
+      elsif @build_state.state == :RUNNING && !response.log
+        raise "Expecting InvocationResponses containing log lines while :RUNNING"
+      elsif (@build_state.state == :BROKEN || @build_state.state == :REJECTED) && !response.error
+        raise "Expecting InvocationResponses containing the error details after :BROKEN or :REJECTED"
+      elsif @build_state.state == :FINISHING && !response.artifact
+        raise "Expecting InvocationResponses containing artifacts while :FINISHING"
+      elsif @build_state.state == :SUCCEEDED || @build_state.state == :FAILED
+        raise "No further InvocationResponse messages expected after :SUCCEEDED or :FAILED"
+      end
+    end
+
+    def process_agent_response(response:)
+      if response.state != :PENDING
+        @build_state.trigger(response.state)
+      else
+        case @build_state.state
+        when :REJECTED
+          # Agent is busy. How do we handle this?
+          emit_error_response(response.error)
+        when :RUNNING
+          emit_new_row(
+            type: response.log.level,
+            message: response.log.message,
+            time: Time.now
+          )
+        when :FINISHING
+          artifact_path = File.join(@artifacts_path, response.artifact.filename)
+          artifact = current_build.artifacts.find { |a| a.reference == artifact_path }
+          unless artifact
+            current_build.artifacts << Artifact.new(
+              type: response.artifact.filename,
+              reference: artifact_path,
+              provider: @project.artifact_provider
+            )
+          end
+          File.open(artifact_path, "a") { |f| f.write(response.artifact.chunk) }
```

Comment:
If I recall correctly, state transition messages do not contain any other sub-messages. Will validate this assumption.